Improve keyboard and screen reader accessibility

dcalhoun opened the pull request:
### Fix
Related to https://github.com/Automattic/simplenote-electron/issues/1935#issuecomment-791742151, this addresses several issue that inhibited keyboard 
and screen reader users from navigation the application easily.

Summary of changes:

- Add ARIA roles for virtualized list
- Improve accessibility of notes list
- Reduce noise in screen reader audio
- Increase accuracy of search field descriptions
- Adjust search button styles to improve tap area
- Increase accuracy of tag list accessibility labels
- Allow scaling to improve accessibility
- Improve ARIA descriptions for application layout
- Add ARIA label for all icon buttons
- Set language for HTML document

| Issue Count Before | Issue Count After |
| --- | --- |
| <img width="1552" alt="axe-devtools-before" src="https://user-images.githubusercontent.com/438664/110390189-38cf7080-802b-11eb-8bf1-b2dbe823fca9.png"> | <img width="1552" alt="axe-devtools-after" src="https://user-images.githubusercontent.com/438664/110391240-d5464280-802c-11eb-9c77-e4ebd28b635b.png"> |

The remaining issues identified by the axe DevTools fall into three categories:

1. Issues that require design input (e.g. text color legibility).
1. Issues with the third-party Monaco editor (e.g. lack of ARIA attributes).
1. Issues that are not applicable to the Simplenote web application (e.g. lack of level-one heading).

### Test
1. Launch Simplenote web or desktop app.
1. Enable macOS VoiceOver or screen reader.
1. Navigate the application.

_Expected Outcome:_ No major issues block navigating the application.

### Release

Improved accessibility of the application for keyboard and screen reader users.

dcalhoun committed:
Add ARIA roles for virtualized list

Elements within `grid` and `rowgroup` require specific roles to be valid
for the ARIA spec.

dcalhoun committed:
Improve accessibility of notes list

- Announce notes list title when focused on list.
- Reduce notes list to one column for easier focus navigation.
- Change actionable elements to buttons to improve accessibility.
- Improve accuracy of note list labels.

dcalhoun committed:
Reduce noise in screen reader audio

This span appears to be unnecessary and was creating an unnecessary
"group" when stepping through UI elements with macOS VoiceOver.

dcalhoun committed:
Increase accuracy of search field descriptions

dcalhoun committed:
Adjust search button styles to improve tap area

Center the icon within the tap area so that the button is easier to tap
with a larger tap area.

dcalhoun committed:
Increase accuracy of tag list accessibility labels

Hide internal text input. Describe available action on tag list.

dcalhoun committed:
Allow scaling to improve accessibility

dcalhoun committed:
Improve ARIA descriptions for application layout

Improve keyboard and screen reader navigation by describing various
pieces of the top-level layout and hide internal developer tooling.

dcalhoun committed:
Add ARIA label for all icon buttons

Ensure screen reader can accurately describe icon buttons that do not
contain text strings.

dcalhoun committed:
Set language for HTML document

dcalhoun commented on lib/search-field/index.tsx:
```diff
@@ -73,21 +73,24 @@ export class SearchField extends Component<Props> {
-    const placeholder = openedTag ?? 'Search notes and tags';
 
-    const screenReaderLabel =
-      'Search ' + (openedTag ? 'notes with tag ' : '') + placeholder;
+    const description =
+      'Search ' +
+      (openedTag ? 'notes with tag ' + openedTag : 'notes and tags');
```
I wanted to draw attention to this change. Previously the placeholder value and the ARIA label were different strings. From reviewing the history of the UI, this at one point made sense because it appears the search field was utilized as the "title" of the list of notes, but also served a search input.

However, several changes were made to this portion of the app — the UI was moved around, the structure of the state, etc. These various changes led to a screen reader description that read "Search Search notes and tags Search notes and tags, search text field." 😂 Needless to say, this description is a bit...confusing.

So, these changes are an attempt to fix this label, but it also involves changing the _visual_ placeholder for the search input. I would appreciate if others would please look carefully at this piece to ensure the current state matches the desired experience.

dcalhoun committed:
Fix cut off icon in macOS Safari

The SVG icon was cut off on the bottom due to its vertical alignment.

dcalhoun committed:
Add additional clarity to dev note

dmsnell commented on lib/app-layout/index.tsx:
```diff
@@ -118,22 +118,28 @@ export class AppLayout extends Component<Props> {
               )}
               {hasRevisions && <RevisionSelector />}
-            </div>
+            </main>
```
nice - I like the changes to `aside` and `main`

dmsnell commented on lib/components/dev-badge/index.tsx:
```diff
@@ -1,7 +1,11 @@
+    <div aria-hidden="true" className="dev-badge">
+      DEV
+    </div>
+  );
```
while we're at it, can we have this return the value directly instead of opening up a function with return here?

```ts
const DevBadge = () => (
	<div aria-hidden="true" className="dev-badge">
		DEV
	</div>
);
```

dmsnell commented on lib/note-list/index.tsx:
```diff
@@ -212,6 +213,19 @@ export class NoteList extends Component<Props> {
+      default:
+        return 'All Notes';
+    }
+  }
```
we don't need to refer to the class instance. can we move this into the module-global scope?

```ts
const sidebarTitle = (collection: Collection) => {
	switch (collection.type) {
		…
	}
}
```

on this I raise two questions:
 - should we extract this function to another file where this module and the list item module can use it together?
 - should we be interpolating the selected tag name here or leave it at "selected tag"?

dcalhoun replied:
Moving it from a class method to the module scope makes sense. ✅ 

> should we extract this function to another file where this module and the list item module can use it together?

I originally avoided doing this because (1) it is only used in two places currently and (2) I hope its [usage here eventually goes away](https://github.com/Automattic/simplenote-electron/blob/0797ce03a44e2493fb2ef06cb5cf0bd958325c13/lib/note-list/index.tsx#L328-L331). That said, I'm happy to abstract it if you'd like. What do you think?

> should we be interpolating the selected tag name here or leave it at "selected tag"?

Yes, it may be a improvement for the title to state the tag name rather than a generic statement. One caveat is that that tag is likely to be visually truncated. Would you like me to change the visual title and ARIA label to be the tag value itself?

<img width="302" alt="Screen Shot 2021-03-09 at 09 05 25" src="https://user-images.githubusercontent.com/438664/110491143-98bf2900-80b6-11eb-8552-c24921ea0954.png">

dmsnell replied:
thanks for the screenshot. now that I'm staring at it "Notes With …" seems confusing. "Notes Tagged …" would seem clearer, but also the truncation looks like it will hit us regularly. I think we used to show the tag name without the additional titling, which didn't have this problem.

this is beyond me at the moment. feel free to reach out to others or make a call for what you think is best.

dcalhoun replied:
>I think we used to show the tag name without the additional titling, which didn't have this problem.

Indeed. I confirmed that. That fact is also related to https://github.com/Automattic/simplenote-electron/pull/2726#discussion_r589807280.

>this is beyond me at the moment. feel free to reach out to others or make a call for what you think is best.

Alright. Hopefully someone else will also find time to review and provide feedback.

belcherj replied:
@SylvesterWilmott what say you?

SylvesterWilmott replied:
The title in the title bar when a tag is selected should be `tag name` to be consistent with `all notes` or `trash`.

Regarding the search placeholder, electron is the only remaining platform where search is not global and so we have this need to imply search is contained only within the selected tag. Once search becomes global (searches all notes all the time) then the search placeholder will read "Search notes and tags". Until then we can try "Search `tag name`" or "Search in `tag name`" to avoid that truncation that looks like it would be a fairly common occurance.

dcalhoun replied:
Thanks for the feedback. I updated the search placeholder to be `Search in <tag_name>`. I also applied an ellipsis when the placeholder overflows the input width.

<img width="309" alt="Screen Shot 2021-03-11 at 11 43 15" src="https://user-images.githubusercontent.com/438664/110830588-fe4e1980-825e-11eb-9ff1-0f0666489e4b.png">

dmsnell commented on lib/note-list/index.tsx:
```diff
@@ -320,6 +339,7 @@ export class NoteList extends Component<Props> {
                     rowHeight={heightCache.rowHeight}
                     rowRenderer={renderNoteRow}
                     scrollToIndex={selectedIndex}
+                    tabIndex={null}
```
what's the difference between `tabIndex={-1}` and `tabIndex={null}`?

dcalhoun replied:
The difference being that `tabIndex={null}` unsets a `tabIndex="0"` [set by default from `react-virtualized`](https://github.com/bvaughn/react-virtualized/issues/275#issuecomment-226077606) and `tabIndex={-1}` disables the element from being focused by keyboard tabbing while still allowing programmatic focus.

Given that the contents of our virtualized list are focusable (buttons to pin or edit a note), IMO it doesn't make sense for the entire list to be a focusable element. We also, to my knowledge, have no need to programmatically focus the list, so `null` makes more sense than `-1`.

dmsnell replied:
interesting. thanks! so the list is no longer focusable: does that have any implications for scrolling the list with the keyboard?

dcalhoun replied:
Good question. I had not explicitly tested scrolling the list with the keyboard until just now. It would appear the scrolling the list is still possible and matches the experience as when `tabindex="0"` is set. 👍🏻

dcalhoun committed:
Refactor for code style preference

dcalhoun committed:
Repair note list styles

Address issues introduced from changing elements from div to button for
better accessibility.

dcalhoun committed:
[wip] Refactor sidebar title computation

Avoid unnecessary class method.

dcalhoun committed:
Adjust search placeholder

Accurately describe that search is scoped to the selected tag. Hide
overflow ellipsis.

dcalhoun committed:
Display selected tag in menu bar

Replace static "Notes With Selected Tag" title with the tag name itself
to provide better context.

sandymcfadden reviewed (approved):
This looks good and works well. Seem much better than what we had previously. Thanks for taking the time to do this.
The only confusing part I had while testing with voiceover was when clicking the button to open the left navigation. Trying to get to the items in it once I clicked the button wasn't straight forward, but I think we can look at that in some future enhancements.

dcalhoun commented:
> The only confusing part I had while testing with voiceover was when clicking the button to open the left navigation. Trying to get to the items in it once I clicked the button wasn't straight forward, but I think we can look at that in some future enhancements.

@sandymcfadden thanks for taking time to review. #2707 _should_ address all issues with navigating the left navigation with a keyboard or screen reader. So, until both PRs are merged, one has to test those areas independently, unfortunately. Please let me know if I am misunderstanding or there are specific issues you encountered.

sandymcfadden commented:
Yes @dcalhoun yeah, of course, you're right. I even looked at and tested the other one. I think I got confused with the two, sorry for that confusion.

codebykat reviewed (approved):
With the caveat that I don't have a lot of domain knowledge here, so in general I trust your expertise. This all looks great!
